crates/energy/src/graph.rs

```diff
@@ -0,0 +1,259 @@
+        }
+    }
+}
+
+/// The energy receiver component is used to mark an entity as an energy receiver.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyReceiver;
+
+/// The energy producer component is used to mark an entity as an energy producer.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyProducer;
+
+/// The nearby component is used to store the nearby entities of an entity.
+#[derive(Debug, Clone, Default)]
+pub struct Nearby {
+    receivers: TinyVec<[NearbyEntity; 32]>,
+    producers: TinyVec<[NearbyEntity; 32]>,
+}
+
+impl Nearby {
+    fn both(&self) -> TinyVec<[NearbyEntity; 64]> {
+        self.receivers
+            .clone()
+            .into_iter()
+            .chain(self.producers.clone())
+            .collect()
+    }
+
+    fn clear(&mut self) {
+        self.receivers.clear();
+        self.producers.clear();
+    }
+
+    fn remove_matching(&mut self, entity: NearbyEntity) {
+        self.receivers.retain(|&e| e != entity);
+        self.producers.retain(|&e| e != entity);
+    }
+}
+
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits {
+    units: Nearby,
+    last_pos: Option<Vec2>,
+}
+
+impl NearbyUnits {
+    pub fn units(&self) -> TinyVec<[NearbyEntity; 64]> {
+        self.units.both()
+    }
+}
+
+fn setup(mut commands: Commands) {
+    commands.insert_resource(PowerGrid::default());
+}
+
+fn clean_up(mut commands: Commands) {
+    commands.remove_resource::<PowerGrid>();
+}
+
+/// This system spawns Energy Producers and Energy Receivers and nearby units.
+fn spawn_graph_components(
+    mut commands: Commands,
+    newly_spawned_units: Query<Entity, Added<Active>>,
+) {
+    for entity in newly_spawned_units.iter() {
+        commands
+            .entity(entity)
+            .insert((EnergyReceiver, NearbyUnits::default()));
+    }
+}
+
+pub fn update_nearby_recv(
+    spacial_index_producer: SpatialQuery<Entity, With<EnergyProducer>>,
+    spacial_index_receiver: SpatialQuery<Entity, With<EnergyReceiver>>,
+    mut nearby_units: Query<(Entity, &mut NearbyUnits, &Transform), Changed<Transform>>,
+) {
+    nearby_units
+        .par_iter_mut()
+        .for_each_mut(|(entity, mut nearby_units, transform)| {
+            if let Some(last_pos) = nearby_units.last_pos {
+                if transform.translation.to_flat().distance_squared(last_pos) < 0.5 {
+                    return;
+                }
+            }
+            nearby_units.last_pos = Some(transform.translation.to_flat());
+
+            let aabb = &Aabb::new(
+                Point::from(transform.translation - Vec3::splat(MAX_DISTANCE)),
+                Point::from(transform.translation + Vec3::splat(MAX_DISTANCE)),
+            );
+
+            let producers = spacial_index_producer.query_aabb(aabb, Some(entity));
+
+            let receivers = spacial_index_receiver.query_aabb(aabb, Some(entity));
+
+            nearby_units.units.clear();
+
+            nearby_units
```

Comment:
I think it might be slightly more efficient to do the spatial query and update the entities before doing the second spatial query (i.e. not having the code interleaved).